<a href="https://colab.research.google.com/github/turna1/3D-LLM/blob/main/Recursion.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Recursion
### by Dr. Rahatara Ferdousi

**OVERVIEW**

 In this notebook, we’ll explore ALL major types of recursion with examples, problems, and visual breakdowns.



# Recursion Types Summary Table

| Recursion Type     | Call Position     | Common Problems                     | Learn More Link |
|--------------------|-------------------|-------------------------------------|-----------------|
| Head Recursion     | Before processing | - Reverse printing<br>- Backtracking | [Backtracking](https://en.wikipedia.org/wiki/Backtracking) |
| Tail Recursion     | After processing  | - Factorial<br>- Sum<br>- Traversal | [Tail Recursion](https://en.wikipedia.org/wiki/Tail_call) |
| Tree Recursion     | Multiple calls    | - Fibonacci<br>- Subsets<br>- Permutations | [Fibonacci](https://en.wikipedia.org/wiki/Fibonacci_number) |
| Indirect Recursion | Between functions | - Alternating logic<br>- Mutual parsing | [Indirect Recursion](https://www.geeksforgeeks.org/indirect-recursion-in-c/) |
| Nested Recursion   | Inside parameters | - McCarthy 91<br>- Ackermann function | [Ackermann Function](https://en.wikipedia.org/wiki/Ackermann_function) |

# Head Recursion
Head recursion defers action until the base case is reached. It's useful when you want to process data in reverse order.

Example: Printing a folder path from deepest to root.


In [ ]:
def print_path_to_folder(folder_list, index=0):
    # Base case: stop when index reaches the end
    if index == len(folder_list):
        return
    # Recursive case: go deeper first
    print_path_to_folder(folder_list, index + 1)
    # Action after recursion: print current folder
    print(folder_list[index], end="/")



In [ ]:
folders = ["root", "home", "user", "documents", "project"]
print_path_to_folder(folders)

project/documents/user/home/root/

Tail recursion performs its action first, then recurses. It's efficient and mimics iteration.
Example: Summing prices in a shopping cart.


In [ ]:
def total_price(cart, index=0, total=0):
    # Base case: all items processed
    if index == len(cart):
        print("Total Price:", total)
        return
    # Recursive case: add current item and move forward (action happen before recursion)
    total_price(cart, index + 1, total + cart[index])



In [ ]:
cart_prices = [19.99, 5.49, 3.50, 12.00]
total_price(cart_prices)

Total Price: 40.98


#Tree recursion
 makes multiple recursive calls. It’s used in problems like generating combinations or Fibonacci numbers.
example: Generating all binary strings of length n.


In [ ]:
def generate_binary(n, s=""):
    # Base case: when no more digits left to add
    if n == 0:
        print(s)
        return

    # Recursive case: branch into '0' and '1'
    generate_binary(n - 1, s + "0")   # Add '0' to current string
    generate_binary(n - 1, s + "1")   # Add '1' to current string



In [ ]:
generate_binary(3)

000
001
010
011
100
101
110
111


```python
generate_binary(3, "")
├── generate_binary(2, "0")         # Add "0" → now s = "0"
│   ├── generate_binary(1, "00")    # Add "0" again → s = "00"
│   │   ├── generate_binary(0, "000") → print "000"  # Base case: print complete string
│   │   └── generate_binary(0, "001") → print "001"  # Base case: print complete string
│   └── generate_binary(1, "01")    # Add "1" → s = "01"
│       ├── generate_binary(0, "010") → print "010"  # Base case
│       └── generate_binary(0, "011") → print "011"  # Base case
└── generate_binary(2, "1")         # Add "1" → now s = "1"
    ├── generate_binary(1, "10")    # Add "0" → s = "10"
    │   ├── generate_binary(0, "100") → print "100"  # Base case
    │   └── generate_binary(0, "101") → print "101"  # Base case
    └── generate_binary(1, "11")    # Add "1" → s = "11"
        ├── generate_binary(0, "110") → print "110"  # Base case
        └── generate_binary(0, "111") → print "111"  # Base case

#Indirect recursion
involves two functions calling each other. It's useful for alternating logic.
Real-world example: Simulating turns in a two-player game.


In [ ]:
def player_one(turns):
    if turns == 0:
        return  # Base case
    print("Player One's turn")
    player_two(turns - 1)  # Recursive case

def player_two(turns):
    if turns == 0:
        return  # Base case
    print("Player Two's turn")
    player_one(turns - 1)  # Recursive case


In [ ]:

player_one(7)

Player One's turn
Player Two's turn
Player One's turn
Player Two's turn
Player One's turn
Player Two's turn
Player One's turn


#Nested Recursion
A function calls itself **multiple times**, branching into several subcalls.  
Classic example: Fibonacci sequence.


In [ ]:
# Recursive function to compute the nth Fibonacci number
def fib(n):
    if n <= 1:
        return n  # Base case: fib(0) = 0, fib(1) = 1

    # Recursive case: sum of previous two Fibonacci numbers
    return fib(n - 1) + fib(n - 2)

# Print Fibonacci numbers from fib(0) to fib(5)
for i in range(6):
    print(f"fib({i}) = {fib(i)}")

fib(0) = 0
fib(1) = 1
fib(2) = 1
fib(3) = 2
fib(4) = 3
fib(5) = 5


```python
fib(5)
├── fib(4)
│   ├── fib(3)
│   │   ├── fib(2)
│   │   │   ├── fib(1) → 1   # base case
│   │   │   └── fib(0) → 0   # base case
│   │   └── fib(1) → 1       # base case
│   └── fib(2)
│       ├── fib(1) → 1       # base case
│       └── fib(0) → 0       # base case
└── fib(3)
    ├── fib(2)
    │   ├── fib(1) → 1       # base case
    │   └── fib(0) → 0       # base case
    └── fib(1) → 1           # base case

Optimising Fibonacci Recursion
* Naïve recursion can be inefficient due to repeated
sub-problems.
* Techniques like memoization and dynamic programming store
results to avoid re-computation.
* These optimisations will be taught in more advanced courses.

### Rule of Thumb: Recursion vs Iteration

- Use **recursion** when the problem naturally breaks into smaller subproblems  
  Examples: tree traversal, backtracking, divide-and-conquer algorithms  
  **Why**: Recursion mirrors the structure of problems that branch or nest — each recursive call handles a smaller piece.

- Use **iteration** when the task is linear, repetitive, or performance-sensitive  
   Examples: summing a list, looping through an array, counting numbers  
  **Why not recursion**: Recursion adds overhead and uses more memory. Iteration is faster and safer for simple, repetitive tasks.


# PARAMETERIZED RECURSION
 We pass parameters that track progress (like counters or sums).
The recursion ends when a condition (base case) stops further calls.




In [ ]:
# ------------------------------------------------------------
# Print a Name N Times
# ------------------------------------------------------------
# Goal: Print the word “Recursion” N times using a recursive function.
# Step 1️ Start with count = 1
# Step 2️ Print message
# Step 3️ Increment count and call function again
# Step 4️ Stop when count > N (base case)


# Visualization:
# print_name(1,3)
# → prints once → calls print_name(2,3)
# → prints again → calls print_name(3,3)
# → prints → calls print_name(4,3) stops

In [ ]:
def print_name(count, n):
    if count > n:  # Base Case
        return
    print(f"{count}. Recursion!")  # Work
    print_name(count + 1, n)       # Recursive Call




In [52]:
print("Example 1️: Print 'Recursion' 5 times")
print_name(1, 7)

Example 1️: Print 'Recursion' 5 times
1. Recursion!
2. Recursion!
3. Recursion!
4. Recursion!
5. Recursion!
6. Recursion!
7. Recursion!


**Goal:** Display numbers both forward and backward using recursion.  
- For 1→N: print first, then recurse forward.  
- For N→1: print first, then recurse backward.  


In [ ]:
# Recursively prints numbers from i to n in ascending order
def print_1_to_n(i, n):
    if i > n:  # Base case: stop recursion when i exceeds n
        return
    print(i, end=" ")  # Print current number
    print_1_to_n(i + 1, n)  # Recursive call with next number




In [ ]:
print("Example: 1 → N")
print_1_to_n(1, 5)

Example: 1 → N
1 2 3 4 5 

In [ ]:
# Recursively prints numbers from n to 1 in descending order
def print_n_to_1(n):
    if n < 1:  # Base case: stop recursion when n drops below 1
        return
    print(n, end=" ")  # Print current number
    print_n_to_1(n - 1)  # Recursive call with previous number


In [ ]:
print("Example: N → 1")
print_n_to_1(5)

Example: N → 1
5 4 3 2 1 

**Goal:** Compute the sum of first N numbers using parameterized recursion.  
Each call carries the running total.


In [ ]:
# Recursively calculates the sum of the first n natural numbers
def sum_n(n, total):
    if n < 1:  # Base case: when n is less than 1, print the final sum
        print("Final Sum:", total)
        return
    sum_n(n - 1, total + n)  # Recursive call with n decremented and total updated



In [ ]:
# Demonstration of the function
print("Example 3: Sum of first 3 numbers")
sum_n(3, 0)  # Output: Final Sum: 6

Example 3: Sum of first 3 numbers
Final Sum: 6


# Functional recursion
 **returns values** instead of carrying them.  
Each call contributes to the final answer during the “unwinding” phase.


In [ ]:
def factorial(n):
    if n == 0:
        return 1
    return n * factorial(n - 1)



In [ ]:
print("Example: Factorial(5) =", factorial(5))


##Backtracking recursion work
 **after** recursion returns.  
Used for reverse-order processing or exploring multiple solutions.


##Problem Statement
Goal: Find a path from the top-left corner to the bottom-right corner of a maze grid.
You can move down or right, and only through cells marked 1 (open path).
Cells marked 0 are walls.
```python
maze = [
    [1, 1, 0],
    [0, 1, 0],
    [0, 1, 1]
]
```

In [ ]:
def solve_maze(maze, row=0, col=0, path=[]):
    rows, cols = len(maze), len(maze[0])

    # Base case: reached bottom-right corner
    if row == rows - 1 and col == cols - 1:
        path.append((row, col))
        print("Path found:", path)
        path.pop()
        return

    # Check boundaries and walls
    if row >= rows or col >= cols or maze[row][col] == 0:
        return

    # Mark current cell as part of path
    path.append((row, col))
    maze[row][col] = 0  # Mark as visited

    # Recursive case: try moving down and right
    solve_maze(maze, row + 1, col, path)  # Move down
    solve_maze(maze, row, col + 1, path)  # Move right

    # Backtrack: undo the move
    maze[row][col] = 1
    path.pop()

In [ ]:
maze = [
    [1, 1, 0],
    [0, 1, 0],
    [0, 1, 1]
]

solve_maze(maze)


Path found: [(0, 0), (0, 1), (1, 1), (2, 1), (2, 2)]


```python
solve_maze(0, 0)
├── move down → solve_maze(1, 0)
│   └── maze[1][0] == 0 → wall → return
└── move right → solve_maze(0, 1)
    ├── move down → solve_maze(1, 1)
    │   ├── move down → solve_maze(2, 1)
    │   │   ├── move down → solve_maze(3, 1) → out of bounds → return
    │   │   └── move right → solve_maze(2, 2)
    │   │       └── ✅ base case → goal reached → print path
    │   └── move right → solve_maze(1, 2) → wall → return
    └── move right → solve_maze(0, 2) → wall → return

[1, 1, 0]   # (0,0) Start → (0,1) → (0,2) ❌ Wall
[0, 1, 0]   # (1,0) ❌ Wall → (1,1) → (1,2) ❌ Wall
[0, 1, 1]   # (2,0) ❌ Wall → (2,1) → (2,2) ✅ Base case
```

**Goal:** Print numbers in ascending order using backtracking.  

**Steps:**
 Recurse to base case (n = 0)  
 Print during the returning phase  


In [ ]:
def print_back(n):
    if n < 1: #base case
        return
    print_back(n - 1)




In [ ]:
# Recursive function to print numbers from 1 to n using backtracking
def print_back(n):
    if n < 1:
        return  # Base case: stop recursion when n is less than 1

    print_back(n - 1)  # Recursive call: go deeper before printing
    print(n, end=" ")  # Backtracking step: print after recursion unwinds



In [ ]:
# Example usage
print("Example: Backtracking 1 → 5")
print_back(5)

Example: Backtracking 1 → 5
1 2 3 4 5 

# SYMMETRIC (TWO-POINTER) RECURSION
Used when processing symmetric structures (start–end).  
Common examples: reversing an array or checking palindromes.

**Goal:** Reverse a list by swapping symmetric elements.  

**Steps:**
- Swap `arr[left]` with `arr[right]`  
- Move inward (`left+1`, `right-1`) until they cross.



In [ ]:
# Recursive function to reverse an array in-place
def reverse_array(arr, left, right):
    if left >= right:
        return  # Base case: pointers have crossed or met, stop recursion

    # Swap the elements at the current left and right indices
    arr[left], arr[right] = arr[right], arr[left]

    # Recursive call: move inward from both ends
    reverse_array(arr, left + 1, right - 1)


In [ ]:


# Example usage
nums = [1, 2, 3, 4, 5]
reverse_array(nums, 0, len(nums) - 1)
print("Example: Reversed array:", nums)

Example: Reversed array: [5, 4, 3, 2, 1]


```python
reverse_array([1, 2, 3, 4, 5], 0, 4)
→ swap arr[0] and arr[4] → [5, 2, 3, 4, 1]
→ reverse_array([5, 2, 3, 4, 1], 1, 3)
   → swap arr[1] and arr[3] → [5, 4, 3, 2, 1]
   → reverse_array([5, 4, 3, 2, 1], 2, 2)
      → left == right → base case → return
```

**Goal:** Check if a word reads the same forward and backward.  
Compare symmetric characters from both ends.


In [ ]:
# Recursive function to check if a string is a palindrome
def is_palindrome(s, i=0):
    # Base case: reached the middle of the string
    if i >= len(s) // 2:
        return True  # All characters matched so far

    # If characters at mirrored positions don't match
    if s[i] != s[-i - 1]:
        return False  # Not a palindrome

    # Recursive case: move inward
    return is_palindrome(s, i + 1)



In [ ]:
# Test cases
for w in ["radar", "level", "queen"]:
    print(f"{w}: {is_palindrome(w)}")

radar: True
level: True
queen: False


```python
i = 0 → s[0] == s[-1] → 'r' == 'r'
i = 1 → s[1] == s[-2] → 'a' == 'a'
i = 2 → middle reached → return True

# Generating Subsequences
Each element can be either **taken** or **skipped**, producing 2ⁿ combinations.
**Goal:** List every possible subsequence of [1, 2, 3].  
At each step:
- Include current element (take)  
- Exclude current element (skip)


In [54]:
def subsequences(i, arr, cur):
    # Base case: if we've considered all elements in the array
    if i == len(arr):
        print(cur)  # Print the current subsequence
        return

    # Recursive case 1: include the current element in the subsequence
    subsequences(i + 1, arr, cur + [arr[i]])  # Take

    # Recursive case 2: exclude the current element from the subsequence
    subsequences(i + 1, arr, cur)             # Skip

In [57]:

subsequences(0, [1, 2], [])


[1, 2]
[1]
[2]
[]


```python
subsequences(0, [1, 2], [])   ← Recursive Call
├── Take 1 → subsequences(1, [1, 2], [1])   ← Recursive Call
│   ├── Take 2 → subsequences(2, [1, 2], [1, 2])   ← Base Case → Print: [1, 2]
│   └── Skip 2 → subsequences(2, [1, 2], [1])      ← Base Case → Print: [1]
└── Skip 1 → subsequences(1, [1, 2], [])   ← Recursive Call
    ├── Take 2 → subsequences(2, [1, 2], [2])      ← Base Case → Print: [2]
    └── Skip 2 → subsequences(2, [1, 2], [])       ← Base Case → Print: []


    subsequences(0, [1, 2], [])
# ├── Take 1 → include arr[0] = 1 → cur becomes [1]
# │   ├── Take 2 → include arr[1] = 2 → cur becomes [1, 2]
# │   │   └── i == 2 → end of array → print [1, 2]
# │   └── Skip 2 → exclude arr[1] → cur remains [1]
# │       └── i == 2 → end of array → print [1]
# └── Skip 1 → exclude arr[0] → cur remains []
#     ├── Take 2 → include arr[1] = 2 → cur becomes [2]
#     │   └── i == 2 → end of array → print [2]
#     └── Skip 2 → exclude arr[1] → cur remains []
#         └── i == 2 → end of array → print []